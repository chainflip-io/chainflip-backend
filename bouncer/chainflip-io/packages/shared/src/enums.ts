type ArrayToMap<T extends readonly string[]> = {
  [K in T[number]]: K;
};

const arrayToMap = <const T extends readonly string[]>(
  array: T,
): ArrayToMap<T> =>
  Object.fromEntries(array.map((key) => [key, key])) as ArrayToMap<T>;

export const Chains = arrayToMap(['Bitcoin', 'Ethereum', 'Polkadot', 'Arbitrum']);
export type Chain = (typeof Chains)[keyof typeof Chains];

export const Assets = arrayToMap(['FLIP', 'USDC', 'DOT', 'ETH', 'BTC', 'ARBETH', 'ARBUSDC']);
export type Asset = (typeof Assets)[keyof typeof Assets];

export const ChainflipNetworks = arrayToMap([
  'sisyphos',
  'perseverance',
  'mainnet',
  'partnernet',
]);

export type ChainflipNetwork =
  (typeof ChainflipNetworks)[keyof typeof ChainflipNetworks];

export const isTestnet = (network: ChainflipNetwork): boolean =>
  network !== ChainflipNetworks.mainnet;

export const assetChains = {
  [Assets.ETH]: Chains.Ethereum,
  [Assets.FLIP]: Chains.Ethereum,
  [Assets.USDC]: Chains.Ethereum,
  [Assets.BTC]: Chains.Bitcoin,
  [Assets.DOT]: Chains.Polkadot,
  [Assets.ARBETH]: Chains.Arbitrum,
  [Assets.ARBUSDC]: Chains.Arbitrum,
} satisfies Record<Asset, Chain>;

export const assetDecimals = {
  [Assets.DOT]: 10,
  [Assets.ETH]: 18,
  [Assets.FLIP]: 18,
  [Assets.USDC]: 6,
  [Assets.BTC]: 8,
  [Assets.ARBETH]: 18,
  [Assets.ARBUSDC]: 6,
} satisfies Record<Asset, number>;

// https://github.com/chainflip-io/chainflip-backend/blob/a2a3c2e447e7b629c4b96797d9eed22eb5b87a0b/state-chain/primitives/src/chains/assets.rs#L51-L59
export const assetContractIds: Record<Asset, number> = {
  // 0 is reservered for particular cross chain messaging scenarios where we want to pass
  // through a message without making a swap.
  [Assets.ETH]: 1,
  [Assets.FLIP]: 2,
  [Assets.USDC]: 3,
  [Assets.DOT]: 4,
  [Assets.BTC]: 5,
  [Assets.ARBETH]: 6,
  [Assets.ARBUSDC]: 7,
};

export const chainAssets = {
  [Chains.Ethereum]: [Assets.ETH, Assets.USDC, Assets.FLIP],
  [Chains.Bitcoin]: [Assets.BTC],
  [Chains.Polkadot]: [Assets.DOT],
  [Chains.Arbitrum]: [Assets.ARBETH, Assets.ARBUSDC],
} satisfies Record<Chain, Asset[]>;

// https://github.com/chainflip-io/chainflip-backend/blob/a2a3c2e447e7b629c4b96797d9eed22eb5b87a0b/state-chain/primitives/src/chains.rs#L52-L56
export const chainContractIds: Record<Chain, number> = {
  [Chains.Ethereum]: 1,
  [Chains.Polkadot]: 2,
  [Chains.Bitcoin]: 3,
  [Chains.Arbitrum]: 4,
};
