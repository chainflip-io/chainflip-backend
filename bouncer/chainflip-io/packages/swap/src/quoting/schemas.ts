import { z } from 'zod';
import { numericString, unsignedInteger } from '@/shared/parsers';

export const marketMakerResponseSchema = z.union([
  z
    .object({
      id: z.string(),
      intermediate_amount: numericString,
      egress_amount: numericString,
    })
    .transform(({ id, ...rest }) => ({
      id,
      intermediateAmount: rest.intermediate_amount,
      egressAmount: rest.egress_amount,
    })),
  z
    .object({
      id: z.string(),
      egress_amount: numericString,
    })
    .transform(({ id, ...rest }) => ({
      id,
      egressAmount: rest.egress_amount,
    })),
]);
export type MarketMakerQuote = z.infer<typeof marketMakerResponseSchema>;

export const swapRateResponseSchema = z
  .object({
    // TODO: simplify when we know how Rust `Option` is encoded
    intermediary: unsignedInteger.optional().nullable(),
    output: unsignedInteger,
  })
  .transform((rate) => ({
    intermediateAmount: rate.intermediary?.toString(),
    egressAmount: rate.output.toString(),
  }));
export type BrokerQuote = z.infer<typeof swapRateResponseSchema>;
