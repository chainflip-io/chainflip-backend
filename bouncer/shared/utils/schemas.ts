import z from 'zod';

export const numericString = z
  .string()
  .regex(/^[\d,]+$/)
  .transform((n) => Number(n.replaceAll(',', '')));

export const hexString = z
  .string()
  .refine((string): string is `0x${string}` => /^0x[a-fA-F0-9]*$/.test(string));
