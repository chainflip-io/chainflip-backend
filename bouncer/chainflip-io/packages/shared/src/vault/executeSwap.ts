import { ContractReceipt } from 'ethers';
import { Vault__factory } from '../abis';
import {
  checkAllowance,
  getTokenContractAddress,
  getVaultManagerContractAddress,
} from '../contracts';
import { assetContractIds, chainContractIds } from '../enums';
import { assert, isTokenSwap } from '../guards';
import {
  executeOptionsSchema,
  type ExecuteOptions,
  type ExecuteSwapParams,
  executeSwapParamsSchema,
  type NativeSwapParams,
  type TokenSwapParams,
} from './schemas';

const swapNative = async (
  { destChain, destAsset, destAddress, amount }: NativeSwapParams,
  { nonce, ...opts }: ExecuteOptions,
): Promise<ContractReceipt> => {
  const vaultContractAddress =
    opts.network === 'localnet'
      ? opts.vaultContractAddress
      : getVaultManagerContractAddress(opts.network);

  const vault = Vault__factory.connect(vaultContractAddress, opts.signer);

  const transaction = await vault.xSwapNative(
    chainContractIds[destChain],
    destAddress,
    assetContractIds[destAsset],
    [],
    { value: amount, nonce },
  );

  return transaction.wait(1);
};

const swapToken = async (
  params: TokenSwapParams,
  opts: ExecuteOptions,
): Promise<ContractReceipt> => {
  const vaultContractAddress =
    opts.network === 'localnet'
      ? opts.vaultContractAddress
      : getVaultManagerContractAddress(opts.network);

  const erc20Address =
    opts.network === 'localnet'
      ? opts.srcTokenContractAddress
      : getTokenContractAddress(params.srcAsset, opts.network);

  assert(erc20Address !== undefined, 'Missing ERC20 contract address');

  const { isAllowable } = await checkAllowance(
    params.amount,
    vaultContractAddress,
    erc20Address,
    opts.signer,
  );
  assert(isAllowable, 'Swap amount exceeds allowance');

  const vault = Vault__factory.connect(vaultContractAddress, opts.signer);

  const transaction = await vault.xSwapToken(
    chainContractIds[params.destChain],
    params.destAddress,
    assetContractIds[params.destAsset],
    erc20Address,
    params.amount,
    [],
    { nonce: opts.nonce },
  );

  return transaction.wait(1);
};

const executeSwap = async (
  params: ExecuteSwapParams,
  options: ExecuteOptions,
): Promise<ContractReceipt> => {
  const parsedParams = executeSwapParamsSchema.parse(params);
  const opts = executeOptionsSchema.parse(options);

  return isTokenSwap(parsedParams)
    ? swapToken(parsedParams, opts)
    : swapNative(parsedParams, opts);
};

export default executeSwap;
