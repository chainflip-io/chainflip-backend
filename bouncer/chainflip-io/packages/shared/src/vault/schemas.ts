import { decodeAddress } from '@polkadot/util-crypto';
import { Signer } from 'ethers';
import { z } from 'zod';
import { Assets, Chains } from '../enums';
import {
  btcAddress,
  chainflipNetwork,
  dotAddress,
  ethereumAddress,
  hexString,
  numericString,
} from '../parsers';

const bytesToHex = (arr: Uint8Array | number[]) =>
  `0x${[...arr].map((v) => v.toString(16).padStart(2, '0')).join('')}`;

const utf8ToHex = (str: string) => `0x${Buffer.from(str).toString('hex')}`;

const base = z.object({
  amount: z.union([numericString, hexString, z.bigint()]),
  srcChain: z.literal(Chains.Ethereum),
  srcAsset: z.literal(Assets.ETH),
});

const ethereumBase = base.extend({
  destChain: z.literal(Chains.Ethereum),
  destAddress: ethereumAddress,
});

const polkadotBase = base.extend({
  destChain: z.literal(Chains.Polkadot),
  destAddress: dotAddress.transform((addr) => bytesToHex(decodeAddress(addr))),
});

const bitcoinBase = base.extend({
  destChain: z.literal(Chains.Bitcoin),
  destAddress: btcAddress.transform(utf8ToHex),
});

const erc20 = z.union([z.literal(Assets.FLIP), z.literal(Assets.USDC)]);

const ethereumNative = ethereumBase.extend({ destAsset: erc20 });

const polkadotNative = polkadotBase.extend({
  destAsset: z.literal(Assets.DOT),
});

const bitcoinNative = bitcoinBase.extend({ destAsset: z.literal(Assets.BTC) });

const nativeSwapParamsSchema = z.union([
  ethereumNative,
  polkadotNative,
  bitcoinNative,
]);

export type NativeSwapParams = z.infer<typeof nativeSwapParamsSchema>;

const tokenSwapParamsSchema = z.union([
  ethereumBase.extend({
    srcAsset: z.literal(Assets.FLIP),
    destAsset: z.union([z.literal(Assets.USDC), z.literal(Assets.ETH)]),
  }),
  ethereumBase.extend({
    srcAsset: z.literal(Assets.USDC),
    destAsset: z.union([z.literal(Assets.FLIP), z.literal(Assets.ETH)]),
  }),
  polkadotNative.extend({ srcAsset: erc20 }),
  bitcoinNative.extend({ srcAsset: erc20 }),
]);

export type TokenSwapParams = z.infer<typeof tokenSwapParamsSchema>;

export const executeSwapParamsSchema = z.union([
  nativeSwapParamsSchema,
  tokenSwapParamsSchema,
]);

export type ExecuteSwapParams = z.infer<typeof executeSwapParamsSchema>;

const nativeCallParamsSchema = ethereumBase.extend({
  destAsset: erc20,
  message: z.string(),
  gasAmount: numericString,
});

export type NativeCallParams = z.infer<typeof nativeCallParamsSchema>;

const tokenCallParamsSchema = z.union([
  ethereumBase.extend({
    srcAsset: z.literal(Assets.FLIP),
    destAsset: z.union([z.literal(Assets.USDC), z.literal(Assets.ETH)]),
    message: hexString,
    gasAmount: numericString,
  }),
  ethereumBase.extend({
    srcAsset: z.literal(Assets.USDC),
    destAsset: z.union([z.literal(Assets.FLIP), z.literal(Assets.ETH)]),
    message: hexString,
    gasAmount: numericString,
  }),
]);

export type TokenCallParams = z.infer<typeof tokenCallParamsSchema>;

export const executeCallParamsSchema = z.union([
  nativeCallParamsSchema,
  tokenCallParamsSchema,
]);

export type ExecuteCallParams = z.infer<typeof executeCallParamsSchema>;

export const executeOptionsSchema = z.intersection(
  z.object({
    signer: z.instanceof(Signer),
    nonce: z.union([z.number(), z.bigint(), z.string()]).optional(),
  }),
  z.union([
    z.object({ network: chainflipNetwork }),
    z.object({
      network: z.literal('localnet'),
      vaultContractAddress: z.string(),
      srcTokenContractAddress: z.string().optional(),
    }),
  ]),
);

export type ExecuteOptions = z.infer<typeof executeOptionsSchema>;
