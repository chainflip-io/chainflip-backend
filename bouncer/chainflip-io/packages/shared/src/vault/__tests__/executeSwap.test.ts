/* eslint-disable @typescript-eslint/no-empty-function */
/* eslint-disable @typescript-eslint/lines-between-class-members */
/* eslint-disable max-classes-per-file */
import { BigNumber, VoidSigner } from 'ethers';
import { Assets, ChainflipNetworks, Chains } from '../../enums';
import executeSwap from '../executeSwap';
import { ExecuteSwapParams } from '../schemas';

const ETH_ADDRESS = '0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2';
const DOT_ADDRESS = '5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX';
const BTC_ADDRESS = 'bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq';

class MockVault {
  constructor(readonly address: string) {}
  async xSwapNative(): Promise<any> {}
  async xSwapToken(): Promise<any> {}
}

class MockERC20 {
  async approve(): Promise<any> {}
  async allowance(): Promise<any> {
    return BigNumber.from(Number.MAX_SAFE_INTEGER - 1);
  }
}

jest.mock('../../abis/factories/Vault__factory', () => ({
  Vault__factory: class {
    static connect: (address: string) => MockVault = jest.fn(
      (address: string) => new MockVault(address),
    );
  },
}));

jest.mock('../../abis/factories/ERC20__factory', () => ({
  ERC20__factory: class {
    static connect: () => MockERC20 = jest.fn(() => new MockERC20());
  },
}));

describe(executeSwap, () => {
  it.each([ChainflipNetworks.perseverance, ChainflipNetworks.mainnet] as const)(
    'only works on sisyphos for now',
    async (network) => {
      await expect(
        executeSwap({} as any, {
          network,
          signer: new VoidSigner('MY ADDRESS'),
        }),
      ).rejects.toThrowError();
    },
  );

  it.each([
    {
      destAsset: Assets.BTC,
      destChain: Chains.Bitcoin,
      destAddress: BTC_ADDRESS,
      srcAsset: Assets.ETH,
      srcChain: Chains.Ethereum,
    },
    {
      destAsset: 'BTC',
      destChain: 'Bitcoin',
      destAddress: BTC_ADDRESS,
      srcAsset: Assets.ETH,
      srcChain: Chains.Ethereum,
    },
    {
      destAsset: Assets.FLIP,
      destChain: Chains.Ethereum,
      destAddress: ETH_ADDRESS,
      srcAsset: Assets.ETH,
      srcChain: Chains.Ethereum,
    },
    {
      destAsset: Assets.USDC,
      destChain: Chains.Ethereum,
      destAddress: ETH_ADDRESS,
      srcAsset: Assets.ETH,
      srcChain: Chains.Ethereum,
    },
    {
      destAsset: Assets.DOT,
      destChain: Chains.Polkadot,
      destAddress: DOT_ADDRESS,
      srcAsset: Assets.ETH,
      srcChain: Chains.Ethereum,
    },
  ] as Omit<ExecuteSwapParams, 'amount'>[])(
    'submits a native swap (%p)',
    async (params) => {
      const wait = jest
        .fn()
        .mockResolvedValue({ status: 1, transactionHash: 'hello world' });
      const swapSpy = jest
        .spyOn(MockVault.prototype, 'xSwapNative')
        .mockResolvedValue({ wait });

      expect(
        await executeSwap({ amount: '1', ...params } as ExecuteSwapParams, {
          network: ChainflipNetworks.sisyphos,
          signer: new VoidSigner('MY ADDRESS'),
        }),
      ).toStrictEqual({ status: 1, transactionHash: 'hello world' });
      expect(wait).toHaveBeenCalledWith(1);
      expect(swapSpy.mock.calls).toMatchSnapshot();
    },
  );

  it.each([
    ...[
      { srcAsset: Assets.FLIP, srcChain: Chains.Ethereum },
      { srcAsset: Assets.USDC, srcChain: Chains.Ethereum },
    ].flatMap((src) => [
      {
        destAsset: Assets.BTC,
        destChain: Chains.Bitcoin,
        destAddress: BTC_ADDRESS,
        ...src,
      },
      {
        destAsset: 'BTC',
        destChain: 'Bitcoin',
        destAddress: BTC_ADDRESS,
        ...src,
      },
      {
        destAsset: Assets.ETH,
        destChain: Chains.Ethereum,
        destAddress: ETH_ADDRESS,
        ...src,
      },
      {
        destAsset: Assets.DOT,
        destChain: Chains.Polkadot,
        destAddress: DOT_ADDRESS,
        ...src,
      },
    ]),
  ] as Omit<ExecuteSwapParams, 'amount'>[])(
    'submits a token swap (%p)',
    async (params) => {
      const wait = jest
        .fn()
        .mockResolvedValue({ status: 1, transactionHash: 'hello world' });
      const approveSpy = jest
        .spyOn(MockERC20.prototype, 'approve')
        .mockResolvedValue({ wait });
      const swapSpy = jest
        .spyOn(MockVault.prototype, 'xSwapToken')
        .mockResolvedValue({ wait });
      const allowanceSpy = jest.spyOn(MockERC20.prototype, 'allowance');

      expect(
        await executeSwap({ amount: '1', ...params } as ExecuteSwapParams, {
          network: 'sisyphos',
          signer: new VoidSigner('MY ADDRESS'),
        }),
      ).toStrictEqual({ status: 1, transactionHash: 'hello world' });
      expect(wait).toHaveBeenCalledWith(1);
      expect(swapSpy.mock.calls).toMatchSnapshot();
      expect(allowanceSpy.mock.calls).toMatchSnapshot();
      expect(approveSpy).not.toHaveBeenCalled();
    },
  );

  it('submits a token swap with sufficient approval', async () => {
    const wait = jest
      .fn()
      .mockResolvedValue({ status: 1, transactionHash: 'hello world' });
    const approveSpy = jest
      .spyOn(MockERC20.prototype, 'approve')
      .mockRejectedValue(Error('unmocked call'));
    const swapSpy = jest
      .spyOn(MockVault.prototype, 'xSwapToken')
      .mockResolvedValue({ wait });
    const allowanceSpy = jest
      .spyOn(MockERC20.prototype, 'allowance')
      .mockResolvedValueOnce(BigNumber.from(Number.MAX_SAFE_INTEGER - 1));

    expect(
      await executeSwap(
        {
          destAsset: Assets.BTC,
          destChain: Chains.Bitcoin,
          destAddress: BTC_ADDRESS,
          srcAsset: Assets.FLIP,
          srcChain: Chains.Ethereum,
          amount: '1',
        },
        { network: 'sisyphos', signer: new VoidSigner('MY ADDRESS') },
      ),
    ).toStrictEqual({ status: 1, transactionHash: 'hello world' });
    expect(wait).toHaveBeenCalledWith(1);
    expect(swapSpy.mock.calls).toMatchSnapshot();
    expect(allowanceSpy.mock.calls).toMatchSnapshot();
    expect(approveSpy).not.toHaveBeenCalled();
  });

  it('can be invoked with localnet options', async () => {
    const wait = jest
      .fn()
      .mockResolvedValue({ status: 1, transactionHash: 'hello world' });
    const approveSpy = jest
      .spyOn(MockERC20.prototype, 'approve')
      .mockResolvedValue({ wait });
    const swapSpy = jest
      .spyOn(MockVault.prototype, 'xSwapToken')
      .mockResolvedValue({ wait });
    const allowanceSpy = jest
      .spyOn(MockERC20.prototype, 'allowance')
      .mockResolvedValueOnce(BigNumber.from(Number.MAX_SAFE_INTEGER - 1));

    expect(
      await executeSwap(
        {
          destAsset: Assets.BTC,
          destChain: Chains.Bitcoin,
          destAddress: BTC_ADDRESS,
          srcAsset: Assets.FLIP,
          amount: '1',
          srcChain: Chains.Ethereum,
        },
        {
          network: 'localnet',
          signer: new VoidSigner('MY ADDRESS'),
          vaultContractAddress: '0x123',
          srcTokenContractAddress: '0x456',
        },
      ),
    ).toStrictEqual({ status: 1, transactionHash: 'hello world' });
    expect(wait).toHaveBeenCalledWith(1);
    expect(swapSpy.mock.calls).toMatchSnapshot();
    expect(allowanceSpy.mock.calls).toMatchSnapshot();
    expect(approveSpy).not.toHaveBeenCalled();
  });

  it.each([1, '1', 1n])('accepts a nonce (%o)', async (nonce) => {
    const wait = jest
      .fn()
      .mockResolvedValue({ status: 1, transactionHash: 'hello world' });
    const swapSpy = jest
      .spyOn(MockVault.prototype, 'xSwapNative')
      .mockResolvedValue({ wait });

    expect(
      await executeSwap(
        {
          amount: '1',
          destAsset: Assets.BTC,
          destChain: Chains.Bitcoin,
          destAddress: BTC_ADDRESS,
          srcAsset: Assets.ETH,
          srcChain: Chains.Ethereum,
        },
        {
          network: ChainflipNetworks.sisyphos,
          signer: new VoidSigner('MY ADDRESS'),
          nonce,
        },
      ),
    ).toStrictEqual({ status: 1, transactionHash: 'hello world' });
    expect(wait).toHaveBeenCalledWith(1);
    expect(swapSpy.mock.calls).toMatchSnapshot();
  });
});
