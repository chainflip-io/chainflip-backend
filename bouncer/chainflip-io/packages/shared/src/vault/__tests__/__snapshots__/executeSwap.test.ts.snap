// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`executeSwap accepts a nonce ('1') 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    [],
    {
      "nonce": "1",
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap accepts a nonce (1) 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    [],
    {
      "nonce": 1,
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap accepts a nonce (1n) 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    [],
    {
      "nonce": 1n,
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap can be invoked with localnet options 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    "0x456",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap can be invoked with localnet options 2`] = `
[
  [
    "MY ADDRESS",
    "0x123",
  ],
]
`;

exports[`executeSwap submits a native swap ({"destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2", "destAsset": "FLIP", "destChain": "Ethereum", "srcAsset": "ETH", "srcChain": "Ethereum"}) 1`] = `
[
  [
    1,
    "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2",
    2,
    [],
    {
      "nonce": undefined,
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap submits a native swap ({"destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2", "destAsset": "USDC", "destChain": "Ethereum", "srcAsset": "ETH", "srcChain": "Ethereum"}) 1`] = `
[
  [
    1,
    "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2",
    3,
    [],
    {
      "nonce": undefined,
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap submits a native swap ({"destAddress": "5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX", "destAsset": "DOT", "destChain": "Polkadot", "srcAsset": "ETH", "srcChain": "Ethereum"}) 1`] = `
[
  [
    2,
    "0x841226ea070c9577979ca2e854130fbe3253853c13c05943e09908312950275d",
    4,
    [],
    {
      "nonce": undefined,
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap submits a native swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin", "srcAsset": "ETH", "srcChain": "Ethereum"}) 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    [],
    {
      "nonce": undefined,
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap submits a native swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin", "srcAsset": "ETH", "srcChain": "Ethereum"}) 2`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    [],
    {
      "nonce": undefined,
      "value": "1",
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2", "destAsset": "ETH", "destChain": "Ethereum", "srcAsset": "FLIP", "srcChain": "Ethereum"}) 1`] = `
[
  [
    1,
    "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2",
    1,
    "0x2FD78122663A07672e9ac63486573A99EB00125d",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2", "destAsset": "ETH", "destChain": "Ethereum", "srcAsset": "FLIP", "srcChain": "Ethereum"}) 2`] = `
[
  [
    "MY ADDRESS",
    "0xd498606dcb9F440CD9F06B397C40fd3327FC2550",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2", "destAsset": "ETH", "destChain": "Ethereum", "srcAsset": "USDC", "srcChain": "Ethereum"}) 1`] = `
[
  [
    1,
    "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2",
    1,
    "0x07865c6E87B9F70255377e024ace6630C1Eaa37F",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2", "destAsset": "ETH", "destChain": "Ethereum", "srcAsset": "USDC", "srcChain": "Ethereum"}) 2`] = `
[
  [
    "MY ADDRESS",
    "0xd498606dcb9F440CD9F06B397C40fd3327FC2550",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX", "destAsset": "DOT", "destChain": "Polkadot", "srcAsset": "FLIP", "srcChain": "Ethereum"}) 1`] = `
[
  [
    2,
    "0x841226ea070c9577979ca2e854130fbe3253853c13c05943e09908312950275d",
    4,
    "0x2FD78122663A07672e9ac63486573A99EB00125d",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX", "destAsset": "DOT", "destChain": "Polkadot", "srcAsset": "FLIP", "srcChain": "Ethereum"}) 2`] = `
[
  [
    "MY ADDRESS",
    "0xd498606dcb9F440CD9F06B397C40fd3327FC2550",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX", "destAsset": "DOT", "destChain": "Polkadot", "srcAsset": "USDC", "srcChain": "Ethereum"}) 1`] = `
[
  [
    2,
    "0x841226ea070c9577979ca2e854130fbe3253853c13c05943e09908312950275d",
    4,
    "0x07865c6E87B9F70255377e024ace6630C1Eaa37F",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "5F3sa2TJAWMqDhXG6jhV4N8ko9SxwGy8TpaNS1repo5EYjQX", "destAsset": "DOT", "destChain": "Polkadot", "srcAsset": "USDC", "srcChain": "Ethereum"}) 2`] = `
[
  [
    "MY ADDRESS",
    "0xd498606dcb9F440CD9F06B397C40fd3327FC2550",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin", "srcAsset": "FLIP", "srcChain": "Ethereum"}) 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    "0x2FD78122663A07672e9ac63486573A99EB00125d",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin", "srcAsset": "FLIP", "srcChain": "Ethereum"}) 2`] = `
[
  [
    "MY ADDRESS",
    "0xd498606dcb9F440CD9F06B397C40fd3327FC2550",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin", "srcAsset": "FLIP", "srcChain": "Ethereum"}) 3`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    "0x2FD78122663A07672e9ac63486573A99EB00125d",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin", "srcAsset": "FLIP", "srcChain": "Ethereum"}) 4`] = `
[
  [
    "MY ADDRESS",
    "0xd498606dcb9F440CD9F06B397C40fd3327FC2550",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin", "srcAsset": "USDC", "srcChain": "Ethereum"}) 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    "0x07865c6E87B9F70255377e024ace6630C1Eaa37F",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin", "srcAsset": "USDC", "srcChain": "Ethereum"}) 2`] = `
[
  [
    "MY ADDRESS",
    "0xd498606dcb9F440CD9F06B397C40fd3327FC2550",
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin", "srcAsset": "USDC", "srcChain": "Ethereum"}) 3`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    "0x07865c6E87B9F70255377e024ace6630C1Eaa37F",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap submits a token swap ({"destAddress": "bc1qar0srrr7xfkvy5l643lydnw9re59gtzzwf5mdq", "destAsset": "BTC", "destChain": "Bitcoin", "srcAsset": "USDC", "srcChain": "Ethereum"}) 4`] = `
[
  [
    "MY ADDRESS",
    "0xd498606dcb9F440CD9F06B397C40fd3327FC2550",
  ],
]
`;

exports[`executeSwap submits a token swap with sufficient approval 1`] = `
[
  [
    3,
    "0x62633171617230737272723778666b7679356c3634336c79646e77397265353967747a7a7766356d6471",
    5,
    "0x2FD78122663A07672e9ac63486573A99EB00125d",
    "1",
    [],
    {
      "nonce": undefined,
    },
  ],
]
`;

exports[`executeSwap submits a token swap with sufficient approval 2`] = `
[
  [
    "MY ADDRESS",
    "0xd498606dcb9F440CD9F06B397C40fd3327FC2550",
  ],
]
`;
